source/common/stats/lazy_init.h

```diff
@@ -0,0 +1,131 @@
+    if (inited_.value() > 0) {
+      inited_.dec();
+    }
+  }
+
+private:
+  inline StatsStructType* instantiate() { return internal_stats_.get(ctor_); }
+
+  // In order to preserve stat value continuity across a config reload, we need to automatically
+  // re-instantiate lazy stats when they are constructed, if there is already a live instantiation
+  // to the same stats. Consider the following alternate scenarios:
+
+  // Scenario 1: a cluster is instantiated but receives no requests, so its traffic-related stats
+  // are never instantiated. When this cluster gets reloaded on a config update, a new lazy-init
+  // block is created, but the stats should again not be instaniated.
+
+  // Scenario 2: a cluster is instantiated and receives traffic, so its traffic-related stats are
+  // instantiated. We must ensure that a new instance for the same cluster gets its lazy-stats
+  // instantiated before the previous cluster of the same name is destructed.
+
+  // To do that we keep an "inited" stat in the cluster's scope, which will be associated by name to
+  // the previous generation's cluster's lazy-init block. We use the value in this shared gauge to
+  // determine whether to instantiate the lazy block on construction.
+  Gauge& inited_;
+  // TODO(stevenzzzz, jmarantz): Clean up this ctor_ by moving its ownership to AtomicPtr, and drop
+  // the setter lambda when the nested object is created.
+  std::function<StatsStructType*()> ctor_;
+  Thread::AtomicPtr<StatsStructType, Thread::AtomicPtrAllocMode::DeleteOnDestruct>
+      internal_stats_{};
+};
+
+// Non-LazyInit wrapper over StatsStructType.
+template <typename StatsStructType>
+class DirectStats : public LazyCompatibleInterface<StatsStructType> {
+public:
+  DirectStats(const typename StatsStructType::StatNameType& stat_names, Stats::Scope& scope)
+      : stats_(stat_names, scope) {}
+
+  // Helper operators to get-or-create and return the StatsStructType object.
+  inline StatsStructType* operator->() override { return &stats_; };
+  inline StatsStructType& operator*() override { return stats_; };
+
+private:
+  StatsStructType stats_;
+};
+
+// A helper class to remove the double indirections on "operator ->/&" for a lazy compatible stats
+// struct.
+template <typename StatsStructType> class LazyCompatibleStats {
+public:
+  static LazyCompatibleStats create(Stats::ScopeSharedPtr scope,
+                                    const typename StatsStructType::StatNameType& stat_names,
+                                    bool lazyinit) {
+    if (lazyinit) {
+      return {std::make_unique<LazyInit<StatsStructType>>(stat_names, scope)};
+    } else {
+      return {std::make_unique<DirectStats<StatsStructType>>(stat_names, *scope)};
+    }
+  }
+
+  // Allows move construct and assign.
+  LazyCompatibleStats& operator=(LazyCompatibleStats&&) = default;
+  LazyCompatibleStats(LazyCompatibleStats&&) = default;
+
+  inline StatsStructType* operator->() { return data_->operator->(); };
```

Comment:
the magic double-indirection is here? this looks good! This is at the edge of my syntactic confidence but if it works it lgtm!

Reply:
yes.
Thanks.